ophyd/status.py

```diff
@@ -1134,3 +1143,299 @@ def wait(status, timeout=None, *, poll_rate="DEPRECATED"):
+                    f"did not reach the desired state {self._operation} {self._value}. "
+                    f"But instead reached {value} in {self._raise_exc_values}, which is set to raise an exception."
+                )
+            )
+            return False
+        return self.OP_MAP[self._operation](value, self._value)
+
+
+class TransitionStatus(SubscriptionStatus):
+    """
+    Status class to monitor transitions of a signal value through a list of specified transitions.
+    The status is finished when all transitions have been observed in order. The keyword argument
+    `strict` determines whether the transitions must occur in strict order or not.
+    If `raise_states` is provided, the status will raise an exception if the signal value matches
+    any of the values in `raise_states`.
+
+    Parameters
+    ----------
+    signal: The device signal to monitor.
+    transitions: A list of values to transition through.
+    strict: Whether the transitions must occur in strict order. Defaults to True.
+    raise_states: A list of values that will raise an exception if encountered. Defaults to None.
+    run: Whether to run the status callback on creation or not. Defaults to True.
+    event_type: The type of event to trigger on transition. Defaults to None (default sub).
+    timeout: The timeout for the status. Defaults to None (indefinite).
+    settle_time: The time to wait for the signal to settle before comparison. Defaults to 0.
+    """
+
+    def __init__(
+        self,
+        signal: Signal,
+        transitions: list[float | int | str],
+        *,
+        strict: bool = True,
+        raise_states: list[float | int | str] | None = None,
+        run: bool = True,
+        event_type=None,
+        timeout: float = None,
+        settle_time: float = 0,
+    ):
+        self._signal = signal
+        if not isinstance(transitions, list):
+            raise ValueError(
+                f"Transitions must be a list of values. Received: {transitions}"
+            )
+        self._transitions = transitions
+        self._index = 0
+        self._strict = strict
+        self._raise_states = raise_states if raise_states else []
+        super().__init__(
+            device=signal,
+            callback=self._compare_callback,
+            timeout=timeout,
+            settle_time=settle_time,
+            event_type=event_type,
+            run=run,
+        )
+
+    def _compare_callback(self, old_value, value, **kwargs) -> bool:
+        """Callback for subscription Status"""
+        if value in self._raise_states:
+            self.set_exception(
+                ValueError(
+                    f"Transition Status for {self._signal.name} resulted in a value: {value}. "
+                    f"marked to raise {self._raise_states}. Expected transitions: {self._transitions}."
+                )
+            )
+            return False
+        if self._index == 0:
+            if value == self._transitions[0]:
```

Comment:
Should this have some checking that the first change we see is to the first state in strict mode?